File: app/src/main/java/com/madebyatomicrobot/things/MainActivity.java
package com.madebyatomicrobot.things;

import android.app.Activity;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.TextView;

import com.google.android.things.pio.Gpio;
import com.google.android.things.pio.GpioCallback;
import com.google.android.things.pio.PeripheralManagerService;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;

// There are lots of better ways of doing this...but it works...
public class MainActivity extends Activity {
    private static final String TAG = MainActivity.class.getSimpleName();

    private TextView primeStatusView;
    private Button startView;
    private Button stopView;

    private Gpio gpio17;
    private Gpio gpio18;

    private AsyncTask<Void, Integer, Void> task;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        primeStatusView = (TextView) findViewById(R.id.primes);
        startView = (Button) findViewById(R.id.start);
        stopView = (Button) findViewById(R.id.stop);

        startView.setEnabled(true);
        stopView.setEnabled(false);

        startView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                showPrimes();
            }
        });

        stopView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                reset();
            }
        });

        PeripheralManagerService manager = new PeripheralManagerService();
        try {
            gpio17 = manager.openGpio("BCM17");
            gpio17.setDirection(Gpio.DIRECTION_OUT_INITIALLY_HIGH);
            gpio17.setActiveType(Gpio.ACTIVE_LOW);

            gpio18 = manager.openGpio("BCM18");
            gpio18.setDirection(Gpio.DIRECTION_IN);
            gpio18.setActiveType(Gpio.ACTIVE_HIGH);

            gpio18.setEdgeTriggerType(Gpio.EDGE_FALLING);
            gpio18.registerGpioCallback(buttonCallback);
        } catch (IOException ex) {
            Log.w(TAG, "Unable to access GPIO", ex);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        cancelTask();

        if (gpio17 != null) {
            try {
                gpio17.close();
                gpio17 = null;
            } catch (IOException e) {
                Log.w(TAG, "Unable to close BCM17", e);
            }
        }

        if (gpio18 != null) {
            gpio18.unregisterGpioCallback(buttonCallback);

            try {
                gpio18.close();
                gpio18 = null;
            } catch (IOException e) {
                Log.w(TAG, "Unable to close BCM18", e);
            }
        }
    }

    private void showPrimes() {
        cancelTask();

        startView.setEnabled(false);
        stopView.setEnabled(true);

        task = new AsyncTask<Void, Integer, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                List<Integer> primes = Arrays.asList(2,3,5,7,11,13,17,19,23,29);
                try {
                    gpio17.setValue(false);
                    for (Integer prime : primes) {
                        publishProgress(prime);
                        for (int i = 0; i < prime; i++) {
                            Thread.sleep(250);
                            gpio17.setValue(true);
                            Thread.sleep(250);
                            gpio17.setValue(false);
                        }

                        gpio17.setValue(false);
                        Thread.sleep(1000);
                    }
                } catch (InterruptedException ex) {
                    return null;
                }
                catch (Exception ex) {
                    Log.e(TAG, "Damn it!", ex);
                }
                return null;
            }

            @Override
            protected void onProgressUpdate(Integer... values) {
                primeStatusView.setText(primeStatusView.getText().toString() + " " + values[0].toString().trim());
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                startView.setEnabled(true);
                stopView.setEnabled(false);

                task = null;
            }
        }.execute();
    }

    private void reset() {
        cancelTask();

        primeStatusView.setText("");
        try {
            gpio17.setValue(false);
        } catch (IOException e) {
            Log.e(TAG, "Damn it!", e);
        }

        startView.setEnabled(true);
        stopView.setEnabled(false);
    }

    private GpioCallback buttonCallback = new GpioCallback() {
        @Override
        public boolean onGpioEdge(Gpio gpio) {
            try {
                if (gpio.getValue()) {
                    Log.v(TAG, "on");
                    showPrimes();
                } else {
                    Log.v(TAG, "off");
                    reset();
                }
            } catch (IOException e) {
                Log.e(TAG, "Damn it!", e);
            }

            return true;
        }

        @Override
        public void onGpioError(Gpio gpio, int error) {
            Log.w(TAG, gpio + ": Error event " + error);
        }
    };

    private void cancelTask() {
        if (task != null) {
            task.cancel(true);
        }
    }
}

File: app/src/main/java/com/madebyatomicrobot/things/MainActivity.kt
package com.madebyatomicrobot.things

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.widget.SeekBar
import android.widget.TextView
import com.google.android.things.pio.I2cDevice
import com.google.android.things.pio.PeripheralManagerService
import java.io.IOException


class MainActivity : Activity() {
    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }

    lateinit var hipYAngleLabel: TextView
    lateinit var hipXAngleLabel: TextView
    lateinit var hipZAngleLabel: TextView
    lateinit var kneeAngleLabel: TextView
    lateinit var ankleAngleLabel: TextView

    lateinit var i2c: I2cDevice

    lateinit var servo03: Servo
    lateinit var servo04: Servo
    lateinit var servo05: Servo
    lateinit var servo06: Servo
    lateinit var servo07: Servo

    lateinit var servo08: Servo
    lateinit var servo09: Servo
    lateinit var servo10: Servo
    lateinit var servo11: Servo
    lateinit var servo12: Servo

    lateinit var leftLeg: Leg
    lateinit var rightLeg: Leg

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        hipYAngleLabel = findViewById(R.id.hip_y_angle_label) as TextView
        hipXAngleLabel = findViewById(R.id.hip_x_angle_label) as TextView
        hipZAngleLabel = findViewById(R.id.hip_z_angle_label) as TextView
        kneeAngleLabel = findViewById(R.id.knee_angle_label) as TextView
        ankleAngleLabel = findViewById(R.id.ankle_angle_label) as TextView

        val manager = PeripheralManagerService()
        try {
            Log.i(TAG, "I2C: " + manager.i2cBusList)
            i2c = manager.openI2cDevice("I2C1", 0x40)

            val pca9685 = PCA9685(i2c)
            pca9685.resetI2C()
            pca9685.setPWMFreq(50.0)  // 50 Hz

            servo03 = Servo(pca9685, 3)
            servo04 = Servo(pca9685, 4)
            servo05 = Servo(pca9685, 5)
            servo06 = Servo(pca9685, 6)
            servo07 = Servo(pca9685, 7)

            servo08 = Servo(pca9685, 8, invertAngle = false)
            servo09 = Servo(pca9685, 9, invertAngle = false)
            servo10 = Servo(pca9685, 10, invertAngle = true, adjustmentAngle = -2.0)
            servo11 = Servo(pca9685, 11, invertAngle = true, adjustmentAngle = 3.0)
            servo12 = Servo(pca9685, 12, invertAngle = true, adjustmentAngle = 5.0)

            leftLeg = Leg(servo07, servo06, servo05, servo04, servo03)
            rightLeg = Leg(servo08, servo09, servo10, servo11, servo12)
        } catch (ex: IOException) {
            Log.w(TAG, "Error in onCreate", ex)
        }

        findViewById(R.id.leg_reset).setOnClickListener { legReset() }
        findViewById(R.id.leg_down).setOnClickListener { legDown() }
        findViewById(R.id.leg_up).setOnClickListener { legUp() }
        findViewById(R.id.left_step).setOnClickListener { leftStep() }
        findViewById(R.id.right_step).setOnClickListener { rightStep() }

        setupSeekBar(R.id.hip_y_angle, Leg.defaultHipYAngle, fun(angle) {
//            servo07.moveToAngle(angle)
//            servo08.moveToAngle(angle)
            leftLeg.moveHipYAngle(angle)
            rightLeg.moveHipYAngle(angle)
        })
        setupSeekBar(R.id.hip_x_angle, Leg.defaultHipXAngle, fun(angle) {
//            servo06.moveToAngle(angle)
//            servo09.moveToAngle(angle)
            leftLeg.moveHipXAngle(angle)
            rightLeg.moveHipXAngle(angle)
        })
        setupSeekBar(R.id.hip_z_angle, Leg.defaultHipZAngle, fun(angle) {
//            servo05.moveToAngle(angle)
//            servo10.moveToAngle(angle)
            leftLeg.moveHipZAngle(angle)
            rightLeg.moveHipZAngle(angle)
        })
        setupSeekBar(R.id.knee_angle, Leg.defaultKneeAngle, fun(angle) {
//            servo04.moveToAngle(angle)
//            servo11.moveToAngle(angle)
            leftLeg.moveKneeAngle(angle)
            rightLeg.moveKneeAngle(angle)
        })
        setupSeekBar(R.id.ankle_angle, Leg.defaultAnkleAngle, fun(angle) {
            //servo03.moveToAngle(angle)
            //servo12.moveToAngle(angle)
            leftLeg.moveAnkleAngle(angle)
            rightLeg.moveAnkleAngle(angle)
        })
    }

    private fun setupSeekBar(
            seekBarResId: Int,
            defaultAngle: Double,
            action: (angle: Double) -> Unit) {
        val seekBar = findViewById(seekBarResId) as SeekBar

        seekBar.max = 180
        seekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                val angle = seekBar!!.progress
                action(angle.toDouble())
                updateLabels()
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
                // Don't care
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {

            }
        })

        seekBar.progress = defaultAngle.toInt()
    }

    private fun updateLabels() {
        hipYAngleLabel.text = "Hip-Y: ${leftLeg.hipYAngle}, ${rightLeg.hipYAngle}"
        hipXAngleLabel.text = "Hip-X: ${leftLeg.hipXAngle}, ${rightLeg.hipXAngle}"
        hipZAngleLabel.text = "Hip-Z: ${leftLeg.hipZAngle}, ${rightLeg.hipZAngle}"
        kneeAngleLabel.text = "Knee: ${leftLeg.kneeAngle}, ${rightLeg.kneeAngle}"
        ankleAngleLabel.text = "Ankle: ${leftLeg.ankleAngle}, ${rightLeg.ankleAngle}"
    }

    override fun onDestroy() {
        try {
            i2c.close()
        } catch (ex: IOException) {
            Log.w(TAG, "Error in onDestroy", ex)
        }

        super.onDestroy()
    }

    fun legReset() {
        leftLeg.reset()
        rightLeg.reset()
        updateLabels()
    }

    fun legDown() {
        leftLeg.down()
        rightLeg.down()
        updateLabels()
    }

    fun legUp() {
        leftLeg.up()
        rightLeg.up()
        updateLabels()
    }

    fun leftStep() {
        Thread({
            leftLeg.adjustHipYAngle(-10.0)
            rightLeg.adjustHipYAngle(-10.0)
            updateLabels()
            Thread.sleep(250)

            rightLeg.up()
            rightLeg.up()
            rightLeg.up()
            rightLeg.up()
            rightLeg.up()
            updateLabels()
            Thread.sleep(250)

            leftLeg.adjustHipXAngle(10.0)
            rightLeg.adjustHipXAngle(10.0)
            updateLabels()
            Thread.sleep(250)

//            leftLeg.down()
//            leftLeg.down()
//            updateLabels()
//            Thread.sleep(250)

            leftLeg.adjustHipYAngle(10.0)
            rightLeg.adjustHipYAngle(10.0)
            updateLabels()
            Thread.sleep(250)

//            leftLeg.adjustHipXAngle(10.0)
//            rightLeg.adjustHipXAngle(10.0)
//            updateLabels()
        }).run()
    }

    fun rightStep() {

    }

    class Leg(val hipY: Servo, val hipX: Servo, val hipZ: Servo, val knee: Servo, val ankle: Servo) {
        companion object {
            val defaultAngleAdjust = 2.0

            val defaultHipYAngle = 90.0
            val defaultHipXAngle = 90.0
            val defaultHipZAngle = 35.0
            val defaultKneeAngle = 70.0
            val defaultAnkleAngle = 45.0
        }

        var hipYAngle = defaultHipYAngle
        var hipXAngle = defaultHipXAngle
        var hipZAngle = defaultHipZAngle
        var kneeAngle = defaultKneeAngle
        var ankleAngle = defaultAnkleAngle

        fun reset() {
            hipYAngle = defaultHipYAngle
            hipXAngle = defaultHipXAngle
            hipZAngle = defaultHipZAngle
            kneeAngle = defaultKneeAngle
            ankleAngle = defaultAnkleAngle
            moveLeg()
        }

        fun up() {
            hipZAngle -= defaultAngleAdjust
            kneeAngle -= defaultAngleAdjust
            ankleAngle += defaultAngleAdjust
            moveLeg()
        }

        fun down() {
            hipZAngle += defaultAngleAdjust
            kneeAngle += defaultAngleAdjust
            ankleAngle -= defaultAngleAdjust
            moveLeg()
        }

        fun adjustHipYAngle(angleAdjust: Double) {
            hipZAngle += angleAdjust
            moveLeg()
        }

        fun adjustHipXAngle(angleAdjust: Double) {
            hipXAngle += angleAdjust
            moveLeg()
        }

        fun moveHipYAngle(angle: Double) {
            hipYAngle = angle
            moveLeg()
        }

        fun moveHipXAngle(angle: Double) {
            hipXAngle = angle
            moveLeg()
        }

        fun moveHipZAngle(angle: Double) {
            hipZAngle = angle
            moveLeg()
        }

        fun moveKneeAngle(angle: Double) {
            kneeAngle = angle
            moveLeg()
        }

        fun moveAnkleAngle(angle: Double) {
            ankleAngle = angle
            moveLeg()
        }

        fun moveLeg() {
            hipY.moveToAngle(hipYAngle)
            hipX.moveToAngle(hipXAngle)
            hipZ.moveToAngle(hipZAngle)
            knee.moveToAngle(kneeAngle)
            ankle.moveToAngle(ankleAngle)
        }
    }
}
